appserver/web/web-glue/src/main/java/com/sun/enterprise/security/web/GlassFishSingleSignOn.java

```diff
@@ -464,44 +461,38 @@ private void processExpires() {
         long tooOld = System.currentTimeMillis() - ssoMaxInactive * 1000L;
         // S1AS8 6155481 START
         if (logger.isLoggable(Level.FINE)) {
-            logger.log(Level.FINE, LogFacade.SSO_EXPIRATION_STARTED, cache.size());
+            logger.log(Level.FINE, LogFacade.SSO_EXPIRATION_STARTED, this.cache.size());
         }
         // S1AS8 6155481 END
-        ArrayList<String> removals = new ArrayList<String>(cache.size() / 2);
+        final ArrayList<String> removals = new ArrayList<>(this.cache.size() / 2);
 
         // build list of removal targets
 
         // Note that only those SSO entries which are NOT associated with
-        // any session are elegible for removal here.
+        // any session are eligible for removal here.
         // Currently no session association ever happens so this covers all
         // SSO entries. However, this should be addressed separately.
 
         try {
-            synchronized (cache) {
-
-                Iterator<String> it = cache.keySet().iterator();
-                while (it.hasNext()) {
-                    String key = it.next();
-                    SingleSignOnEntry sso = (SingleSignOnEntry) cache.get(key);
-                    if (sso.isEmpty() && sso.getLastAccessTime() < tooOld) {
-                        removals.add(key);
-                    }
+            this.cache.forEach((ssoId, sso) -> {
+                if (sso.isEmpty() && sso.getLastAccessTime() < tooOld) {
+                    removals.add(ssoId);
                 }
```

Comment:
There is a little change in behavior. 

Old implementation locked SSOs and created a list of expired SSOs. No one could create a new SSO.

New implementation uses what is available and does not block creation of a new SSO (which is not taken for expiration in gicen round). However probability of a new SSO being expired is so small so I think this change should be OK for a real world usage.